smack-websocket/src/main/java/org/jivesoftware/smack/websocket/XmppWebsocketTransportModule.java

```diff
@@ -0,0 +1,221 @@
+import org.jivesoftware.smack.XMPPException;
+import org.jivesoftware.smack.c2s.ModularXmppClientToServerConnection.ConnectedButUnauthenticatedStateDescriptor;
+import org.jivesoftware.smack.c2s.ModularXmppClientToServerConnection.LookupRemoteConnectionEndpointsStateDescriptor;
+import org.jivesoftware.smack.c2s.ModularXmppClientToServerConnectionModule;
+import org.jivesoftware.smack.c2s.XmppClientToServerTransport;
+import org.jivesoftware.smack.c2s.internal.ModularXmppClientToServerConnectionInternal;
+import org.jivesoftware.smack.c2s.internal.WalkStateGraphContext;
+import org.jivesoftware.smack.fsm.State;
+import org.jivesoftware.smack.fsm.StateDescriptor;
+import org.jivesoftware.smack.fsm.StateTransitionResult;
+import org.jivesoftware.smack.fsm.StateTransitionResult.AttemptResult;
+import org.jivesoftware.smack.packet.TopLevelStreamElement;
+import org.jivesoftware.smack.tcp.XmppTcpTransportModule.EstablishingTcpConnectionStateDescriptor;
+import org.jivesoftware.smack.util.rce.RemoteConnectionEndpointLookupFailure;
+import org.jivesoftware.smack.websocket.XmppWebsocketTransportModule.XmppWebsocketTransport.DiscoveredWebsocketEndpoints;
+import org.jivesoftware.smack.websocket.elements.OpenElement;
+import org.jivesoftware.smack.websocket.rce.RemoteWebsocketConnectionEndpoints;
+import org.jivesoftware.smack.websocket.rce.RemoteWebsocketConnectionEndpoints.Result;
+import org.jivesoftware.smack.websocket.rce.WebsocketRemoteConnectionEndpoint;
+
+public final class XmppWebsocketTransportModule
+                extends ModularXmppClientToServerConnectionModule<XmppWebsocketTransportModuleDescriptor> {
+    private final XmppWebsocketTransport websocketTransport;
+
+    private WebsocketConnectionAttemptState connectionAttemptState;
+
+    protected XmppWebsocketTransportModule(XmppWebsocketTransportModuleDescriptor moduleDescriptor,
+                    ModularXmppClientToServerConnectionInternal connectionInternal) {
+        super(moduleDescriptor, connectionInternal);
+
+        websocketTransport = new XmppWebsocketTransport(connectionInternal);
+    }
+
+    private void afterOutgoingElementsQueueModified() {
+        TopLevelStreamElement topLevelStreamElement = connectionInternal.outgoingElementsQueue.poll();
+        connectionAttemptState.websocket.send(topLevelStreamElement);
+    }
+
+    @Override
+    protected XmppWebsocketTransport getTransport() {
+        return websocketTransport;
+    }
+
+    static final class EstablishingWebsocketConnectionStateDescriptor extends StateDescriptor {
+        private EstablishingWebsocketConnectionStateDescriptor() {
+            super(XmppWebsocketTransportModule.EstablishingWebsocketConnectionState.class);
+            addPredeccessor(LookupRemoteConnectionEndpointsStateDescriptor.class);
+            addSuccessor(ConnectedButUnauthenticatedStateDescriptor.class);
+            declareInferiortyTo(EstablishingTcpConnectionStateDescriptor.class);
```

Comment:
It should not be possible that the `smack-websocket` subproject has access to the `smack-tcp` classes. This should use the, not yet existent, String based methods to declare inferiority and precedence over other states. But this is not part of this milestone's goals (something for the third phase).